app/models/device.rb

```diff
@@ -17,4 +17,20 @@ def pretty_owner
   def two_digit_location
     ISO3166::Country.find_country_by_alpha3(self.location).alpha2
   end
+
+  def config_attributes
+    self.attributes.filter do |attr|
+      !%w(uuid user_id created_at updated_at tags display_name group).include?(attr)
+    end
+  end
+
+  def determine_group
+    group_devices = Device.group(:group).where.not(uuid: self.uuid)
+    group_devices.each do |device|
+      if self.config_attributes == device.config_attributes
+        return device.group
+      end
+    end
+    Device.pluck(:group).max.to_i + 1
```

Comment:
This throws an error if there are devices with `nil` for `group`. I know in theory only the device to be saved should be missing a group but idk what would happen if we got multiple entries at once. Also this would be a problem if we want to update the entries we've got in the db currently (as they'll all have nil group)

Reply:
That's a good point, I'm going to look into it. I'll at least make a workaround for the error, although it would be nice if it also assigned groups to existing nodes without them.

Reply:
I think assigning groups to existing devices would be easiest with a data migration - https://github.com/ilyakatz/data-migrate

(in future)

Reply:
Most recent commit prevents the error, but doesn't include any data migration. Will mark that on the list of Future Features.

In the meantime, any devices with `nil` groups are grouped together and their rank on the grouped leaderboard is based on the specs of the oldest device. This should only affect devices created before this PR, everything from here on should never have a `nil` group.